Add a safe version of to_date(), which validates the input date

andreasscherbaum opened the pull request:
The regular to_date() function accepts invalid dates and uses
the input to calculate a valid date based on the difference in
the input. 2016-02-31 becomes 2016-03-02.

The new function verifies that the output and the input date
are the same, using the format string provides as second
parameter. A difference raises an error.

Solves Tracker #119368939

ivannovick commented:
@csylvester-pivotal can you help review this functionality?

ascherbaum-pivotal commented:
The conflict is from the big 8.3 merge. Will resolve before commit.

csylvester-pivotal commented:
Thanks Andreas. This is a welcomed addition to GPDB and will solve an issue raised by a US Fed program I support.

andreasscherbaum commented:
@csylvester-pivotal Thanks for the review. Any change requests in this PR, or is it good to go?

csylvester-pivotal commented:
@andreasscherbaum Looks good to me. I don't have anything to add.

hlinnaka commented on src/backend/utils/adt/formatting.c:
```diff
@@ -3306,6 +3306,53 @@ to_date(PG_FUNCTION_ARGS)
+	/* call the original to_date() function and receive the result */
+	result = DirectFunctionCall2(to_date, PG_GETARG_POINTER(0), PG_GETARG_POINTER(1));
+	if ((Pointer *)result == NULL)
+	{
```
I don't think this can happen, to_date() never returns NULL

hlinnaka replied:
DateADT is not a pointer type anyway, so comparing it with NULL is bogus

andreasscherbaum replied:
Hmm, you ar right. I was looking at another function first, that's why I added this guard. Will remove it.

hlinnaka commented on src/backend/utils/adt/formatting.c:
```diff
@@ -3306,6 +3306,53 @@ to_date(PG_FUNCTION_ARGS)
+		{
+			ereport(ERROR,
+					(errcode(ERRCODE_DATETIME_VALUE_OUT_OF_RANGE),
+					 errmsg("date out of range: \"%s\"",
```
It could be invalid in many other ways besides "out of range", right?

andreasscherbaum replied:
Right, but most of it should be handled by to_date(). If anything comes to your mind which is not handled by to_date(), will return a date from to_date() and is still invalid, please let me know.

hlinnaka commented on src/test/regress/parallel_schedule:
```diff
@@ -141,3 +141,6 @@ test: default_parameters
 # 'zlib' utilizes fault injectors so it needs to be in a group by itself
 test: zlib
+
+# test to_date() and to_date_valid()
```
Put this to greenplum_schedule instead. parallel_schedule is inherited from upstream, and we're trying to keep it unchanged from the upstream version, to avoid merge conflicts.

hlinnaka commented on src/test/regress/expected/to_date.out:
```diff
@@ -0,0 +1,96 @@
+   DateStyle   
+---------------
+ Postgres, MDY
+(1 row)
```
Are these tests sensitive to datestyle? Perhaps it'd be better to explicitly set the right datestyle, than just check that it is what we expect.

andreasscherbaum replied:
I thought about it, because my first tests failed. That's why I added this test, instead of changing the setting. It will fail if someone changes the default setting.

My opinion is that this is a better indicator for trouble.

hlinnaka commented on src/backend/utils/adt/formatting.c:
```diff
@@ -3306,6 +3306,53 @@ to_date(PG_FUNCTION_ARGS)
+	/* compare the two dates */
+	validate = DirectFunctionCall2(timestamp_to_char, TimestampGetDatum(validate_in), PG_GETARG_POINTER(1));
+	if ((Pointer *)validate == NULL)
+	{
```
DirectFunctionCall throws an error if the function returns NULL.

hlinnaka commented on src/backend/utils/adt/formatting.c:
```diff
@@ -3306,6 +3306,53 @@ to_date(PG_FUNCTION_ARGS)
+/* ----------
+ * TO_DATE_VALID
+ *	Make Date from date_str which is formated at argument 'fmt'
+ *	Validate the input date afterwards
```
Would be good to explain briefly how the validation is done. I.e. convert back and compare

andreasscherbaum replied:
Added an additional note

andreasscherbaum commented:
@hlinnaka I now moved the tests into the Greenplum schedule, copied the DirectFunctionCallX functionality for NULL handling and cleaned up all warnings. Also added a few more tests, because it looks like to_date() has no tests at all in the existing regression tests.

Can you please have a look if that is ok?

Thanks

hlinnaka commented on src/backend/utils/adt/formatting.c:
```diff
@@ -3297,6 +3298,61 @@ to_date(PG_FUNCTION_ARGS)
+	fcinfo2.argnull[1] = false;
+
+	validate = timestamp_to_char(&fcinfo2);
+	if ((Pointer *)validate == NULL)
```
The correct way to check for NULL result is to check fcinfo2.isnull. Timestamp is a typedef for int64, so casting it to Pointer is bogus (0 is a valid value).

hlinnaka commented on src/backend/utils/adt/formatting.c:
```diff
@@ -3297,6 +3298,61 @@ to_date(PG_FUNCTION_ARGS)
+	}
+	else
+	{
+		if (strcmp(text_to_cstring((text *)validate), text_to_cstring(date_txt)) != 0)
```
While this works, note that text_to_cstring() makes a palloc'd copy of the input. That might be acceptable, but I gather that we're treating this function is at least somewhat performance critical - otherwise we would just have implemented it in PL/pgSQL. Perhaps it'd be better to call texteq(), which avoids those copies?

A quick micro-benchmark between this and regular to_date() might be in order..

andreasscherbaum replied:
Timings:

strcmp/text_to_cstring: 10000 tests, 5m45.980s, 5m52.349s, 5m52.113s
textne: 10000 tests, 5m45.086s, 5m51.155s, 5m52.756s

There is no difference. I assume that using texteq/textne and the required DirectFunctionCall2() also adds some overhead. However this solution looks nicer, so let's go with this one.

hlinnaka commented on src/backend/utils/adt/formatting.c:
```diff
@@ -3297,6 +3298,61 @@ to_date(PG_FUNCTION_ARGS)
+Datum
+to_date_valid(PG_FUNCTION_ARGS)
+{
+	text		*date_txt = PG_GETARG_TEXT_P(0);
```
Note that PG_GETARG_TEXT_P() de-toasts the input if it's toasted. You're actually only using date_txt in the error messages, so that's a waste of effort unless there's an error.

The usual style is to call PG_GETARG_* macros in the beginning of the function, like you did here for date_txt, so I would leave that as it is. In fact, I'd call add a variable and use a PG_GETARG_* macro for the 2nd argument too, for consistency, and replace the PG_GETARG_DATUM() invocations below with the variables. When those macros are in the very beginning of the function, it's easy to see from the source what the arguments to it are, as that's not otherwise immediately clear for functions that use the PG_FUNCTION_ARG mechanism.

You could save some cycles by using PG_GETARG_TEXT_PP() instead of PG_GETARG_TEXT_P(), though, as that avoids making a copy of the input if it's a "packed" varlen with 1-byte header.

andreasscherbaum replied:
I'm also using date_txt in the strcmp(), so I have to de-toast it anyway.

hlinnaka commented:
Do we need a to_timestamp_safe() variant too?

andreasscherbaum commented:
@ivannovick That is a question for you: do we need a to_timestamp_valid() function as well?

ivannovick commented:
does to_timestamp() have the ability to give "unexpected" results with bad inputs?

andreasscherbaum commented:
@ivannovick to_timestamp() has the same problem:
```
postgres=# SELECT to_timestamp('2014/02/27', 'YYYY/MM/DD'); 
      to_timestamp      
------------------------
 2014-02-27 00:00:00+01
(1 row)

postgres=# SELECT to_timestamp('2014/02/31', 'YYYY/MM/DD'); 
      to_timestamp      
------------------------
 2014-03-03 00:00:00+01
(1 row)
```

ivannovick commented:
@andreasscherbaum I see.  We definitely need the to_timestamp_safe() variant in this case.

hlinnaka commented on src/backend/utils/adt/formatting.c:
```diff
@@ -3297,6 +3298,62 @@ to_date(PG_FUNCTION_ARGS)
+	InitFunctionCallInfoData(fcinfo2, NULL, 2, NULL, NULL);
+
+	fcinfo2.arg[0] = validate_in;
+	fcinfo2.arg[1] = PG_GETARG_DATUM(1);
```
Would be good to use 'fmt' here, otherwise timestamp_to_char() needs to detoast it again.

hlinnaka commented on src/backend/utils/adt/formatting.c:
```diff
@@ -3297,6 +3298,62 @@ to_date(PG_FUNCTION_ARGS)
+	fcinfo2.argnull[0] = false;
+	fcinfo2.argnull[1] = false;
+
+	validate = timestamp_to_char(&fcinfo2);
```
On closer look, this NULL handling still looks wrong:

* Can timestamp_to_char() ever actually return a NULL? There's a PG_RETURN_NULL() in there, but it only returns NULL if datetime_to_char_body() returns NULL, and I don't think it ever does.

* Assuming that timestamp_to_char() can return NULL anyway, would returning NULL from to_date_valid() be correct behaviour? It would be different from what to_date() does for the same input, and would be rather surprising IMHO.

I'm thinking that this should go back to a straight DirectFunctionCall2(), after all..

hlinnaka replied:
Ok, there's another PG_RETURN_NULL() at the top of in timestamp_to_char() that I didn't see when I wrote the above. But the 2nd point remains: is returning NULL from to_date_valid() a good idea?

hlinnaka replied:
Might be worth adding an explicit TIMESTAMP_NOT_FINITE() check into to_date_valid(), before calling timestamp_to_char(). I'm not sure if to_date() can ever return a non-finite timestamp, but the raison d'etre for this function is to be sure you don't get values you didn't expect into your database, so better safe than sorry.

andreasscherbaum commented:
ok, all changes from Heikki are in.

Off to to_timestamp_valid() ...

andreasscherbaum commented:
@ivannovick to_timestamp() doesn't really work this way. The timezone handling can and will produce different dates than the original input - and I don't see a way to verify between "different date" and "different date because of invalid input".

ivannovick commented:
@andreasscherbaum what was the latest on this?  Is it fixed upstream and are we planning to merge this PR?

ascherbaum-pivotal commented:
Upstream has a bigger discussion going on about changing this, together with several proposals. I'll have an eye on it, but so far nothing is committed.

ivannovick commented:
ok lets close for now
